library/core/src/ptr/mod.rs

```diff
@@ -65,6 +65,8 @@
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
 //!
+//! It is guaranteed that an allocated object never spans more than `isize::MAX` bytes.
```

Comment:
```suggestion
//! It is guaranteed that an allocated object never spans more than `isize::MAX` bytes.
//! An allocated object can never contain the null address, and the address "at the end"
//! (i.e., one past the last byte that is contained inside the allocated object) can always
//! be computed without wrapping around the address space. (The last part is equivalent
//! to saying that `usize::MAX` is never contained in an allocated object.)
```

Reply:
Should we be worried that "null" isn't well-defined? It sounds like it might be up in the air whether "null" always means 0?

Reply:
We should probably say somewhere that null is 0x00, but I'm not sure where.

Reply:
Perhaps we can just leave null out for now? It'd be trivial to add later, and I'm not aware of any code that needs to anchor on that property (there are other existing language constructs that provide this guarantee - e.g. the existence of [`From<&T> for NonNull<T>`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html#impl-From%3C%26T%3E-for-NonNull%3CT%3E)).

Reply:
I feel like if we're stating general allocation properties anyway, this one is worth stating, even if we don't define the representation of a null ptr yet. But I can do that in a follow-up PR if you prefer. It might just mean going through FCP twice.^^

Reply:
Maybe something like the following so people don't rely on the null address being zero?

> An allocated object can never contain the null address. While the null address is usually the address with the numerical value 0, this is not currently guaranteed to be the case. Until this is guaranteed, unsafe code which assumes that the null address is 0 may be unsound on some target architectures or under future versions of Rust.

Reply:
Honestly I would prefer this to explicitly say that null = 0, and then we nominate for t-lang discussion and we'll see whether they think that's already clear or FCP-worthy.

But if you don't want to do that, then it's probably better to not mention the null stuff at all. I will then do it in a follow-up PR. (We'll need to nominate for t-lang discussion either way, though.)

Reply:
Submitted https://github.com/rust-lang/rust/issues/116895 to track this. I'm not opposed to landing that first and then updating the text here and landing this PR second, assuming that process is relatively quick.

Reply:
Nah let's just do this separately, I don't want to hold up your PR.